site/app/controllers/admin/UsersController.php

```diff
@@ -1000,14 +937,30 @@ public function uploadUserList($list_type = "classlist") {
                     $bad_columns[] = 'user_email';
                 }
             }
-            /* $row[4] validation varies by $list_type
-               "classlist" validates registration_section, and "graderlist" validates user_group */
-            if (!$row4_validation_function()) {
-                $bad_row_details[$row_num + 1][] = $list_type === 'classlist'
-                    ? 'Registration section'
-                    : 'Grader-group';
-                if (!in_array('row4_validation', $bad_columns)) {
-                    $bad_columns[] = 'row4_validation';
+            /* Check registration for appropriate format. Allowed characters - A-Z,a-z,_,- .
+            Registration section is optional for graders, so automatically validate if not set.*/
+            if (isset($vals[$registration_section_idx]) && strtolower($vals[$registration_section_idx]) === "null") {
+                $vals[$registration_section_idx] = null;
+                $uploaded_data[$row_num][$registration_section_idx] = null;
```

Comment:
So once again, this is the 1 line change for bug fix.

Explanation: changes to `$vals` (which is an alias to a row in `$uploaded_data`) doesn't reflect in the actual `$uploaded_data`. So later, when we do (I am abbreviating the logic in line 1070 of `UserController.php`):
`foreach ($uploaded_data as $row) {
      ...
      $this->core->getQueries()->insertNewRegistrationSection($row[$registration_section_idx]);
      ...
}`
it still inserts unchanged value, a.k.a. "NULL" string, into the database.

Reply:
Looking at the code, it looks like we set a value of `$vals` a second time during this loop, probably with the intention that'll stick till later?

It may be better to adjust the foreach to be:

```
foreach ($uploaded_data as $row_num => $vals) {
    // ... existing code
    $uploaded_data[$row_num] = $vals;
}
```

This ensures we update the row for `$uploaded_data` after manipulating it, without having to update both arrays each time.

An alternative would be to pass-by-reference in the foreach itself:
```
foreach ($uploaded_data as $row_num => &$vals) {
    // ... rest of the code
}
unset($vals);
```

where we pass-by-reference the value for `$vals` and so changing it will change the original data. I prefer to avoid this as requiring the `unset` after the `foreach` is somewhat unexpected for new people (see warning on https://www.php.net/manual/en/control-structures.foreach.php), and it's just a bit of a pitfall versus the above, and I don't think the data is so large we need to save on memory here.

Reply:
Good suggestion, and thanks for sharing about the pass-by-reference didn't know that was possible! I will request for your re-review once I test just to be safe.